Allow individual int values per option

danielappelt opened the pull request:
This allows each element in a Chooser to be mapped to its own special integer.

A new constructor allows to supply an extra array of int values which are used for this purpose. When updating the Chooser from the model, the element with the next best available value is selected.

eclab commented:
I have a revised version of Chooser which does this but it's not tested.  It's very close to your approach.  Contact me directly (email is in manual) and I'll get it to you.

Sean

> On Apr 14, 2019, at 5:11 PM, Daniel Appelt <notifications@github.com> wrote:
> 
> This allows each element in a Chooser to be mapped to its own special integer.
> 
> A new constructor allows to supply an extra array of int values which are used for this purpose. When updating the Chooser from the model, the element with the next best available value is selected.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/eclab/edisyn/pull/13
> 
> Commit Summary
> 
> 	• Allow individual int values per option
> File Changes
> 
> 	• M edisyn/gui/Chooser.java (78)
> Patch Links:
> 
> 	• https://github.com/eclab/edisyn/pull/13.patch
> 	• https://github.com/eclab/edisyn/pull/13.diff
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

danielappelt commented:
Will do!

eclab commented:
Daniel, I have committed a new version of Edisyn with a few new patch editors.  I believe it has changes to the Chooser in-line with what you were proposing, but it's been quite a while since I looked at that and my memory is fuzzy.  Before I announced the posted version, could you check to see if the Chooser is to your liking, and if not, re-clue me in on what to tweak?

Sean

> On Apr 14, 2019, at 11:07 PM, Sean Luke <sean@cs.gmu.edu> wrote:
> 
> I have a revised version of Chooser which does this but it's not tested.  It's very close to your approach.  Contact me directly (email is in manual) and I'll get it to you.
> 
> Sean
> 
>> On Apr 14, 2019, at 5:11 PM, Daniel Appelt <notifications@github.com> wrote:
>> 
>> This allows each element in a Chooser to be mapped to its own special integer.
>> 
>> A new constructor allows to supply an extra array of int values which are used for this purpose. When updating the Chooser from the model, the element with the next best available value is selected.
>> 
>> You can view, comment on, or merge this pull request online at:
>> 
>>  https://github.com/eclab/edisyn/pull/13
>> 
>> Commit Summary
>> 
>> 	• Allow individual int values per option
>> File Changes
>> 
>> 	• M edisyn/gui/Chooser.java (78)
>> Patch Links:
>> 
>> 	• https://github.com/eclab/edisyn/pull/13.patch
>> 	• https://github.com/eclab/edisyn/pull/13.diff
>> —
>> You are receiving this because you are subscribed to this thread.
>> Reply to this email directly, view it on GitHub, or mute the thread.
>> 
>

danielappelt commented:
Thanks Sean! I recently moved and therefore did not have much time to spend on this.

There are a few things that I think are still missing. For example, in the update method the model's value is still coupled closely to the number of elements in the Chooser (lines 77-80). I will take a closer look at this tomorrow.

Finally, I come more and more to the conclusion that it might not be a good idea to use the same key in edisyn for equally keyed parameters of a sysex spec which have very different value ranges.

danielappelt commented:
I now returned to your original version of the update method and just corrected the bits where state should get restricted to be between the minimum and maximum allowed values. Furthermore, I added a method to get the selected value.

This version of the Chooser should behave identical to your implementation if no separate values array is used and implement an identical behaviour if values are provided. This version also works for my current editor. So, this would be my final proposal for now.

danielappelt commented:
Hey Sean,

some time ago, I updated the original pull request. Could you please check
whether this version would also work for you?:
https://github.com/eclab/edisyn/pull/13/files

Thank you!
Daniel

Am Mi., 12. Juni 2019 um 19:23 Uhr schrieb Evolutionary Computation
Laboratory <notifications@github.com>:

> Daniel, I have committed a new version of Edisyn with a few new patch
> editors. I believe it has changes to the Chooser in-line with what you were
> proposing, but it's been quite a while since I looked at that and my memory
> is fuzzy. Before I announced the posted version, could you check to see if
> the Chooser is to your liking, and if not, re-clue me in on what to tweak?
>
> Sean
>
> > On Apr 14, 2019, at 11:07 PM, Sean Luke <sean@cs.gmu.edu> wrote:
> >
> > I have a revised version of Chooser which does this but it's not tested.
> It's very close to your approach. Contact me directly (email is in manual)
> and I'll get it to you.
> >
> > Sean
> >
> >> On Apr 14, 2019, at 5:11 PM, Daniel Appelt <notifications@github.com>
> wrote:
> >>
> >> This allows each element in a Chooser to be mapped to its own special
> integer.
> >>
> >> A new constructor allows to supply an extra array of int values which
> are used for this purpose. When updating the Chooser from the model, the
> element with the next best available value is selected.
> >>
> >> You can view, comment on, or merge this pull request online at:
> >>
> >> https://github.com/eclab/edisyn/pull/13
> >>
> >> Commit Summary
> >>
> >> • Allow individual int values per option
> >> File Changes
> >>
> >> • M edisyn/gui/Chooser.java (78)
> >> Patch Links:
> >>
> >> • https://github.com/eclab/edisyn/pull/13.patch
> >> • https://github.com/eclab/edisyn/pull/13.diff
> >> —
> >> You are receiving this because you are subscribed to this thread.
> >> Reply to this email directly, view it on GitHub, or mute the thread.
> >>
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/eclab/edisyn/pull/13?email_source=notifications&email_token=AAX6CWBUHRYQN4N3SPNLKDDP2EWJ3A5CNFSM4HF3NA7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXRF5GI#issuecomment-501374617>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAX6CWENBPUWF4U47O2YGOTP2EWJ3ANCNFSM4HF3NA7A>
> .
>

eclab commented:
Dan, the pull request you submitted before isn't compatible with the current version of Chooser.  So to get it merged in for you I have to do one of three things:

	A] Figure out how to update it to the current Chooser version
	B] Write an equivalent version
	C] Revert Chooser, merge it in, and then push forward with fixes since then

I think you could do [A], but if I was doing it, [B] would be easiest.  So to do [B] to start, I'm trying to remember what your objective was.  I think it was that you needed to be able to map Chooser list values to arbitrary numbers.  I think I had implemented that somewhat differently than your pull request, but we weren't sure mine worked right and your implementation had some problem that I cannot remember right now.  IIRC, ultimately you said you'd had second thoughts about it and had shelved the your project, so I think I went ahead and made Chooser simpler again.

SO: let me try to get a version working for you.  Can you remind me exactly what you need Chooser to do that it can't do, and let me try committing a new version.

Sean

danielappelt commented:
Hey! Over here, Github is saying that the current version of the pull request has no conflicts with your master branch. It could be merged as is.

What I want to achieve with this pull request is to be able to provide an arbitrary list of values instead of just using [0..<number of elements - 1>] as Chooser values.

Your current version in master has this in place for the construction of the Chooser (except for line 290). Apart from this, the pull request suggests to check against the minimum and maximum available values when updating the Chooser's value. If you don't provide a separate list of values during construction, the minimum will be 0 and the maximum will be <number of elements - 1>. So, this solution just falls back exactly to the behaviour that you have in place right now if you don't provide a separate list of values during construction.

Finally, I suggest to add a method getValue() in order to get access to the currently selected value (cf. getIndex()).

eclab commented:
As I was working on this, it occurred to me that there's a showstopper problem which may doom the whole thing.  :-(

The problem is that Edisyn's model assumes that parameters have an integer numerical range from a...b and that it is contiguous -- there are no holes in it.  NumericalComponents (such as Chooser) normally obey this.  This is critical for Edisyn's randomization algorithms to work right (mutation, merging, hill-climbing, constricting, nudging).  And the vast majority of synths operate in this manner.  Some synths may have a *non-metric* range and a *metric* range for a given parameter -- for example, a MIDI channel might be 1-16 (the metric part) plus 0=OFF, 17=OMNI, 18=MPE_LOWER, 19=MPE_UPPER -- and Edisyn is smart about handling that -- but it's still a continguous range 0...19.

The problem with the Chooser proposal is that it makes possible storage of *non-contiguous* ranges in the model, and this will break the randomization tools which are a hallmark of Edisyn.  These algorithms don't know or care that it's a Chooser (say) in control of a given parameter -- just that a parameter has a min, max, metric min, and metric max.

I think what this means is that this shouldn't be done at the Chooser level at all.  It should be simplified to just require fixed numerical ranges.

So how would we handle this?  It'd need to be handled at the sysex parameter parse/emit level.  That is, when reading a sysex parameter or whatnot, we'd map it to the contiguous space and put it in the model, and when emitting a parameter, we'd map it back.

To make this easier, I'd suggest a standard simple utility mapping class, or alternatively adding a mapping  facility to the model (though that'd be confusing I think and wouldn't add anything).  I could take a stab at something simple.

BTW, remind me what synth you're writing an editor for, and which parameter is causing the issue.

Sean

> On Aug 7, 2019, at 3:31 PM, Daniel Appelt <notifications@github.com> wrote:
> 
> Hey! Over here, Github is saying that the current version of the pull request has no conflicts with your master branch. It could be merged as is.
> 
> What I want to achieve with this pull request is to be able to provide an arbitrary list of values instead of just using [0..<number of elements - 1>] as Chooser values.
> 
> Your current version in master has this in place for the construction of the Chooser (except for line 290). Apart from this, the pull request suggests to check against the minimum and maximum available values when updating the Chooser's value. If you don't provide a separate list of values during construction, the minimum will be 0 and the maximum will be <number of elements - 1>. So, this solution just falls back exactly to the behaviour that you have in place right now if you don't provide a separate list of values during construction.
> 
> Finally, I suggest to add a method getValue() in order to get access to the currently selected value (cf. getIndex()).
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
